### PR TITLE
Feature/ccclxxxvi 8

### PR DESCRIPTION
**Pull request by VolodymyrBodnar**
## JIRA

* CCCLXXXVI-8

## Code reviewers

- [ ] @Andriy97091
- [ ] @irynazaiets
- [ ] @jelysaw
- [ ] @MaksymGromyk
- [ ] @shado1997
- [ ] @tavor118
- [ ] @VolodymyrBodnar
- [ ] @wm139

### Second Level Review

- [ ] @lhalam
- [ ] @alutsysh

## Summary of issue

Edit expend
## Summary of change

Added forms for editing expend

## Testing approach

unittests with nose

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by irynazaiets — DISMISSED** [review]

**Review by tavor118 — DISMISSED** [review]

File: moneta/src/python/db/expend.py [hunk 1]
```diff
@@ -1,37 +1,116 @@
...
+        "this method execute transaction query via pool_manager"
         with DBPoolManager().get_cursor() as curs:
-            # Execute the SQL command
             curs.execute(query, args)
 
     @staticmethod
-    def __get_from_db(query):
-        '''Connecting to database and reterning result of SELECT.'''
+    def __get_from_db(query, args):
+        "this method execute query and return some record from db as tuple of tuples"
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: Not propper doc string

File: moneta/src/python/db/expend.py [hunk 2]
```diff
@@ -1,37 +1,116 @@
...
-            cursor.execute(query)
-            result = cursor.fetchall()
-        return result
+            curs = conn.cursor(DictCursor)
+            curs.execute(query, args)
+            return curs.fetchall()
+
+    @staticmethod
+    def edit_name(expend_id, new_name):
+        """method for renaming expend"""
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: Method

File: moneta/src/python/db/expend.py [hunk 3]
```diff
@@ -1,37 +1,116 @@
...
+        args = (new_name, expend_id,)
+        Expend.__execute_query(query, args)
+
+    @staticmethod
+    def edit_amount(expend_id, new_amount):
+        """method for editing planned cost in expend"""
+        try:
+            new_amount = int(new_amount)
+        except ValueError:
+            pass
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: what reason for this check?

File: moneta/www/views/expend.py [hunk 4]
```diff
@@ -1,9 +1,107 @@
...
+
+@login_required
+def expend_main(request):
+    "view which shows list of all user's expends"
+
+    current_user = request.user
+    if current_user.id:
+        user_id = current_user.id
+    else:
+        user_id = 1
```
Comment (alutsysh) [inline moneta/www/views/expend.py]: WHY?

**Review by MaksymGromyk — DISMISSED** [review]

**Review by irynazaiets — DISMISSED** [review]

**Review by jelysaw — DISMISSED** [review]

**Review by tavor118 — DISMISSED** [review]

**Review by irynazaiets — DISMISSED** [review]

**Review by tavor118 — DISMISSED** [review]

**Review by vengvas — DISMISSED** [review]

**Review by irynazaiets — DISMISSED** [review]

**Review by vengvas — DISMISSED** [review]

**Review by tavor118 — DISMISSED** [review]

File: moneta/src/python/db/expend.py [hunk 5]
```diff
@@ -1,37 +1,132 @@
...
+        if new_name:
+            query_args.append('name = %s')
+            args.append(new_name)
+        if new_amount:
+            query_args.append('amount = %s')
+            args.append(new_amount)
+        if new_image_id:
+            query_args.append('image_id = %s')
+            args.append(new_image_id)
+        query = 'UPDATE expend SET ' + ','.join(query_args) + ' WHERE id = %s;'
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: no +! use query = 'UPDATE expend SET %s WHERE %%s;' % (','.join(query_args),)

File: moneta/src/python/db/expend.py [hunk 6]
```diff
@@ -1,37 +1,132 @@
...
+        args.append(expend_id)
+        args = tuple(args)
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: I think this is a useless transformation

File: moneta/src/python/db/expend.py [hunk 7]
```diff
@@ -1,37 +1,132 @@
...
+        """
+        Check if user has permission to edit expend record.
+
+        Args:
+            user_id(int): id of logged user.
+            expend_id(int) : id of expend.
+        Returns:
+             True or False.
+        """
+        query = """
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: Use one style for query

File: moneta/src/python/db/expend.py [hunk 8]
```diff
@@ -1,37 +1,133 @@
...
+        if new_amount:
+            query_args.append('amount = %s')
+            args.append(new_amount)
+        if new_image_id:
+            query_args.append('image_id = %s')
+            args.append(new_image_id)
+
+        query = 'UPDATE expend SET %s WHERE id = %s;' % (','.join(query_args), expend_id)
+        Expend._make_transaction(query, args)
+        LOGGER.info('expend %s was updated with query %s, %s.' % (expend_id, query, args))
```
Comment (alutsysh) [inline moneta/src/python/db/expend.py]: LOGGER.info('expend %s was updated with query %s, %s.', expend_id, query, args)

File: moneta/www/views/expend.py [hunk 9]
```diff
@@ -1,13 +1,130 @@
...
+        raise PermissionDenied()
+
+    if request.method == 'POST':
+        form = EditExpendForm(request.POST)
+        if form.is_valid():
+            new_name = form.cleaned_data.get('new_name')
+            new_amount = form.cleaned_data.get('new_amount')
+            new_image = form.cleaned_data.get('new_image')
+            Expend.update(expend_id, new_name, new_amount, new_image)
+            LOGGER.info('user %s update expend %s', (request.user.id, expend_id))
```
Comment (alutsysh) [inline moneta/www/views/expend.py]: LOGGER.info('user %s update expend %s', request.user.id, expend_id)

**Review by jelysaw — APPROVED** [review]

**Review by alutsysh — APPROVED** [review]

**Review by MaksymGromyk — APPROVED** [review]

**Review by tavor118 — APPROVED** [review]

**Pull request merged**